File: macos-hud/JARVIS-HUD/UniversalWebSocketClient.swift
import Foundation
import Combine

/// Universal WebSocket Client for JARVIS
/// Supports dynamic configuration, health checks, capability negotiation, and state synchronization
/// Works with any backend WebSocket server following the JARVIS protocol v2.0
class UniversalWebSocketClient: NSObject, ObservableObject {

    // MARK: - Configuration

    struct Configuration: Codable {
        let version: String
        struct WebSocketConfig: Codable {
            let url: String
            let ready: Bool
            let healthCheckRequired: Bool
            struct ReconnectConfig: Codable {
                let enabled: Bool
                let maxAttempts: Int
                let baseDelayMs: Int
                let maxDelayMs: Int
                let jitter: Bool

                enum CodingKeys: String, CodingKey {
                    case enabled
                    case maxAttempts = "max_attempts"
                    case baseDelayMs = "base_delay_ms"
                    case maxDelayMs = "max_delay_ms"
                    case jitter
                }
            }
            let reconnect: ReconnectConfig

            enum CodingKeys: String, CodingKey {
                case url, ready
                case healthCheckRequired = "health_check_required"
                case reconnect
            }
        }
        struct HTTPConfig: Codable {
            let baseUrl: String
            let healthEndpoint: String
            let configEndpoint: String

            enum CodingKeys: String, CodingKey {
                case baseUrl = "base_url"
                case healthEndpoint = "health_endpoint"
                case configEndpoint = "config_endpoint"
            }
        }
        let websocket: WebSocketConfig
        let http: HTTPConfig
        let capabilities: [String]
    }

    struct HealthCheckResponse: Codable {
        let status: String
        let websocketReady: Bool
        let version: String
        let uptime: Double
        let capabilities: [String]

        enum CodingKeys: String, CodingKey {
            case status
            case websocketReady = "websocket_ready"
            case version, uptime, capabilities
        }
    }

    struct WelcomeMessage: Codable {
        let type: String
        let message: String
        let serverVersion: String
        let protocolVersion: String
        let capabilities: [String]
        let bufferedMessagesCount: Int

        enum CodingKeys: String, CodingKey {
            case type, message
            case serverVersion = "server_version"
            case protocolVersion = "protocol_version"
            case capabilities
            case bufferedMessagesCount = "buffered_messages_count"
        }
    }

    // MARK: - Published Properties

    @Published var connectionState: ConnectionState = .disconnected
    @Published var serverVersion: String = "unknown"
    @Published var serverCapabilities: [String] = []
    @Published var lastError: String?

    enum ConnectionState {
        case disconnected
        case fetchingConfig
        case healthCheck
        case connecting
        case connected
        case reconnecting(attempt: Int)
        case error(String)
    }

    // MARK: - Private Properties

    private var websocketTask: URLSessionWebSocketTask?
    private var configuration: Configuration?
    private var reconnectAttempt = 0
    private var reconnectTimer: Timer?
    private var messageHandler: ((String) -> Void)?

    private let clientId: String
    private let clientType: String
    private let clientCapabilities: [String]
    private let defaultBaseURL: String

    // MARK: - Initialization

    init(
        clientId: String = UUID().uuidString,
        clientType: String = "hud",
        capabilities: [String] = ["hud_client", "voice", "vision"],
        baseURL: String = "http://localhost:8010"
    ) {
        self.clientId = clientId
        self.clientType = clientType
        self.clientCapabilities = capabilities
        self.defaultBaseURL = baseURL
        super.init()
    }

    // MARK: - Public API

    /// Connect to the WebSocket server with full protocol flow
    func connect(messageHandler: @escaping (String) -> Void) {
        self.messageHandler = messageHandler

        Task {
            await performConnectionFlow()
        }
    }

    /// Disconnect from WebSocket
    func disconnect() {
        reconnectTimer?.invalidate()
        reconnectTimer = nil

        websocketTask?.cancel(with: .goingAway, reason: nil)
        websocketTask = nil

        DispatchQueue.main.async {
            self.connectionState = .disconnected
        }
    }

    /// Send a message to the server
    func send(_ message: [String: Any]) {
        guard let websocketTask = websocketTask else {
            print("❌ Cannot send message: WebSocket not connected")
            return
        }

        do {
            let jsonData = try JSONSerialization.data(withJSONObject: message)
            let jsonString = String(data: jsonData, encoding: .utf8) ?? ""
            let message = URLSessionWebSocketTask.Message.string(jsonString)

            websocketTask.send(message) { error in
                if let error = error {
                    print("❌ WebSocket send error: \(error)")
                }
            }
        } catch {
            print("❌ Failed to serialize message: \(error)")
        }
    }

    // MARK: - Connection Flow

    private func performConnectionFlow() async {
        print("🚀 Starting Universal WebSocket Connection Flow...")

        // Step 1: Fetch configuration
        await fetchConfiguration()

        guard let config = configuration else {
            await setError("Failed to fetch configuration")
            return
        }

        // Step 2: Health check (if required)
        if config.websocket.healthCheckRequired {
            let healthy = await performHealthCheck(baseURL: config.http.baseUrl)
            if !healthy {
                await setError("Health check failed - backend not ready")
                await scheduleReconnect()
                return
            }
        }

        // Step 3: Connect to WebSocket
        await connectToWebSocket(url: config.websocket.url)
    }

    private func fetchConfiguration() async {
        await MainActor.run {
            connectionState = .fetchingConfig
        }

        print("📡 Fetching configuration from backend...")

        let configURL = URL(string: "\(defaultBaseURL)/api/config")!

        do {
            let (data, response) = try await URLSession.shared.data(from: configURL)

            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                print("❌ Config endpoint returned non-200 status")
                // Fallback to defaults
                await useFallbackConfiguration()
                return
            }

            let decoder = JSONDecoder()
            let config = try decoder.decode(Configuration.self, from: data)

            self.configuration = config
            print("✅ Configuration fetched successfully")
            print("   WebSocket URL: \(config.websocket.url)")
            print("   Server capabilities: \(config.capabilities.joined(separator: ", "))")

        } catch {
            print("⚠️  Failed to fetch config: \(error)")
            print("   Using fallback configuration...")
            await useFallbackConfiguration()
        }
    }

    private func useFallbackConfiguration() async {
        // Create fallback config with defaults
        let fallbackConfig = Configuration(
            version: "2.0.0",
            websocket: Configuration.WebSocketConfig(
                url: "ws://localhost:8010/ws",
                ready: true,
                healthCheckRequired: true,
                reconnect: Configuration.WebSocketConfig.ReconnectConfig(
                    enabled: true,
                    maxAttempts: 999,
                    baseDelayMs: 500,
                    maxDelayMs: 15000,
                    jitter: true
                )
            ),
            http: Configuration.HTTPConfig(
                baseUrl: defaultBaseURL,
                healthEndpoint: "/health",
                configEndpoint: "/api/config"
            ),
            capabilities: ["voice", "vision", "commands"]
        )

        self.configuration = fallbackConfig
        print("✅ Fallback configuration loaded")
    }

    private func performHealthCheck(baseURL: String) async -> Bool {
        await MainActor.run {
            connectionState = .healthCheck
        }

        print("🏥 Performing health check...")

        let healthURL = URL(string: "\(baseURL)/health")!

        do {
            let (data, response) = try await URLSession.shared.data(from: healthURL)

            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                print("❌ Health check failed: non-200 status")
                return false
            }

            let decoder = JSONDecoder()
            let health = try decoder.decode(HealthCheckResponse.self, from: data)

            print("✅ Health check passed")
            print("   Status: \(health.status)")
            print("   WebSocket Ready: \(health.websocketReady)")
            print("   Version: \(health.version)")
            print("   Uptime: \(Int(health.uptime))s")

            await MainActor.run {
                serverVersion = health.version
                serverCapabilities = health.capabilities
            }

            return health.websocketReady && (health.status == "healthy" || health.status == "degraded")

        } catch {
            print("❌ Health check error: \(error)")
            return false
        }
    }

    private func connectToWebSocket(url: String) async {
        await MainActor.run {
            connectionState = .connecting
        }

        print("🔌 Connecting to WebSocket: \(url)")

        guard let websocketURL = URL(string: url) else {
            await setError("Invalid WebSocket URL")
            return
        }

        let session = URLSession(configuration: .default)
        let task = session.webSocketTask(with: websocketURL)

        self.websocketTask = task
        task.resume()

        // Send capability negotiation handshake
        await sendHandshake()

        // Start receiving messages
        await receiveMessages()

        await MainActor.run {
            connectionState = .connected
            reconnectAttempt = 0
        }

        print("✅ WebSocket connected successfully")
    }

    private func sendHandshake() async {
        let handshake: [String: Any] = [
            "type": "client_connect",
            "client_id": clientId,
            "client_type": clientType,
            "version": "2.0.0",
            "capabilities": clientCapabilities,
            "request_state": true
        ]

        send(handshake)
        print("📤 Sent capability negotiation handshake")
    }

    private func receiveMessages() async {
        guard let task = websocketTask else { return }

        do {
            let message = try await task.receive()

            switch message {
            case .string(let text):
                handleMessage(text)

            case .data(let data):
                if let text = String(data: data, encoding: .utf8) {
                    handleMessage(text)
                }

            @unknown default:
                print("⚠️  Unknown message type")
            }

            // Continue receiving
            await receiveMessages()

        } catch {
            print("❌ WebSocket receive error: \(error)")
            await handleDisconnection()
        }
    }

    private func handleMessage(_ text: String) {
        // Parse message to detect welcome
        if let data = text.data(using: .utf8),
           let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let type = json["type"] as? String,
           type == "welcome" {

            // Extract server info from welcome message
            if let decoder = try? JSONDecoder(),
               let welcomeData = try? JSONDecoder().decode(WelcomeMessage.self, from: data) {
                DispatchQueue.main.async {
                    self.serverVersion = welcomeData.serverVersion
                    self.serverCapabilities = welcomeData.capabilities
                }
                print("📨 Received welcome message")
                print("   Server version: \(welcomeData.serverVersion)")
                print("   Protocol version: \(welcomeData.protocolVersion)")
                print("   Buffered messages: \(welcomeData.bufferedMessagesCount)")
            }
        }

        // Forward to message handler
        messageHandler?(text)
    }

    private func handleDisconnection() async {
        print("🔌 WebSocket disconnected")

        websocketTask = nil

        await scheduleReconnect()
    }

    private func scheduleReconnect() async {
        guard let config = configuration,
              config.websocket.reconnect.enabled else {
            await setError("Reconnection disabled")
            return
        }

        reconnectAttempt += 1

        guard reconnectAttempt <= config.websocket.reconnect.maxAttempts else {
            await setError("Max reconnection attempts reached")
            return
        }

        await MainActor.run {
            connectionState = .reconnecting(attempt: reconnectAttempt)
        }

        // Calculate backoff delay
        let baseDelay = Double(config.websocket.reconnect.baseDelayMs) / 1000.0
        let maxDelay = Double(config.websocket.reconnect.maxDelayMs) / 1000.0

        var delay = min(baseDelay * pow(2.0, Double(reconnectAttempt - 1)), maxDelay)

        // Add jitter
        if config.websocket.reconnect.jitter {
            let jitter = Double.random(in: -0.1...0.1) * delay
            delay += jitter
        }

        print("🔄 Reconnecting in \(String(format: "%.1f", delay))s (attempt \(reconnectAttempt)/\(config.websocket.reconnect.maxAttempts))")

        try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))

        await performConnectionFlow()
    }

    private func setError(_ message: String) async {
        await MainActor.run {
            connectionState = .error(message)
            lastError = message
        }
        print("❌ \(message)")
    }
}

File: frontend/src/utils/UniversalWebSocketClient.js
/**
 * Universal WebSocket Client for JARVIS
 *
 * Supports dynamic configuration, health checks, capability negotiation, and state synchronization
 * Works with any backend WebSocket server following the JARVIS protocol v2.0
 *
 * Features:
 * - Zero hardcoded URLs (uses /api/config endpoint)
 * - Health check before WebSocket connection
 * - Automatic reconnection with exponential backoff
 * - Capability negotiation
 * - State synchronization
 * - Message buffering and replay
 *
 * @example
 * const client = new UniversalWebSocketClient({
 *   clientType: 'web',
 *   capabilities: ['voice', 'vision'],
 *   baseURL: 'http://localhost:8010' // Optional, will auto-detect
 * });
 *
 * client.on('message', (data) => console.log('Received:', data));
 * client.on('stateChange', (state) => console.log('State:', state));
 *
 * await client.connect();
 */

class UniversalWebSocketClient {
  constructor(options = {}) {
    // Client configuration
    this.clientId = options.clientId || this.generateClientId();
    this.clientType = options.clientType || 'web';
    this.clientCapabilities = options.capabilities || ['voice'];
    this.baseURL = options.baseURL || this.detectBaseURL();

    // Connection state
    this.connectionState = 'disconnected';
    this.websocket = null;
    this.configuration = null;
    this.reconnectAttempt = 0;
    this.reconnectTimer = null;

    // Server info
    this.serverVersion = 'unknown';
    this.serverCapabilities = [];

    // Event handlers
    this.eventHandlers = {
      message: [],
      stateChange: [],
      error: [],
      connected: [],
      disconnected: []
    };

    // Message queue for when disconnected
    this.messageQueue = [];

    console.log('🚀 UniversalWebSocketClient initialized', {
      clientId: this.clientId,
      clientType: this.clientType,
      capabilities: this.clientCapabilities,
      baseURL: this.baseURL
    });
  }

  // MARK: - Public API

  /**
   * Connect to the WebSocket server with full protocol flow
   */
  async connect() {
    try {
      await this.performConnectionFlow();
    } catch (error) {
      console.error('❌ Connection flow failed:', error);
      this.emit('error', error.message);
      await this.scheduleReconnect();
    }
  }

  /**
   * Disconnect from WebSocket
   */
  disconnect() {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }

    if (this.websocket) {
      this.websocket.close(1000, 'Client disconnect');
      this.websocket = null;
    }

    this.setState('disconnected');
  }

  /**
   * Send a message to the server
   */
  send(message) {
    if (!this.websocket || this.websocket.readyState !== WebSocket.OPEN) {
      console.warn('⚠️  WebSocket not connected, queueing message');
      this.messageQueue.push(message);
      return false;
    }

    try {
      this.websocket.send(JSON.stringify(message));
      return true;
    } catch (error) {
      console.error('❌ Failed to send message:', error);
      this.emit('error', `Send failed: ${error.message}`);
      return false;
    }
  }

  /**
   * Add event listener
   */
  on(event, handler) {
    if (this.eventHandlers[event]) {
      this.eventHandlers[event].push(handler);
    }
  }

  /**
   * Remove event listener
   */
  off(event, handler) {
    if (this.eventHandlers[event]) {
      this.eventHandlers[event] = this.eventHandlers[event].filter(h => h !== handler);
    }
  }

  // MARK: - Connection Flow

  async performConnectionFlow() {
    console.log('🚀 Starting Universal WebSocket Connection Flow...');

    // Step 1: Fetch configuration
    this.setState('fetchingConfig');
    await this.fetchConfiguration();

    if (!this.configuration) {
      throw new Error('Failed to fetch configuration');
    }

    // Step 2: Health check (if required)
    if (this.configuration.websocket.health_check_required) {
      this.setState('healthCheck');
      const healthy = await this.performHealthCheck();
      if (!healthy) {
        throw new Error('Health check failed - backend not ready');
      }
    }

    // Step 3: Connect to WebSocket
    this.setState('connecting');
    await this.connectToWebSocket();
  }

  async fetchConfiguration() {
    console.log('📡 Fetching configuration from backend...');

    try {
      const response = await fetch(`${this.baseURL}/api/config`);

      if (!response.ok) {
        throw new Error(`Config endpoint returned ${response.status}`);
      }

      const config = await response.json();
      this.configuration = config;

      console.log('✅ Configuration fetched successfully');
      console.log('   WebSocket URL:', config.websocket.url);
      console.log('   Server capabilities:', config.capabilities.join(', '));

    } catch (error) {
      console.warn('⚠️  Failed to fetch config:', error);
      console.log('   Using fallback configuration...');
      this.useFallbackConfiguration();
    }
  }

  useFallbackConfiguration() {
    const wsProtocol = this.baseURL.startsWith('https') ? 'wss:' : 'ws:';
    const wsURL = this.baseURL.replace(/^https?:/, wsProtocol) + '/ws';

    this.configuration = {
      version: '2.0.0',
      websocket: {
        url: wsURL,
        ready: true,
        health_check_required: true,
        reconnect: {
          enabled: true,
          max_attempts: 999,
          base_delay_ms: 500,
          max_delay_ms: 15000,
          jitter: true
        }
      },
      http: {
        base_url: this.baseURL,
        health_endpoint: '/health',
        config_endpoint: '/api/config'
      },
      capabilities: ['voice', 'vision', 'commands']
    };

    console.log('✅ Fallback configuration loaded');
  }

  async performHealthCheck() {
    console.log('🏥 Performing health check...');

    try {
      const healthURL = `${this.configuration.http.base_url}/health`;
      const response = await fetch(healthURL);

      if (!response.ok) {
        throw new Error(`Health check returned ${response.status}`);
      }

      const health = await response.json();

      console.log('✅ Health check passed');
      console.log('   Status:', health.status);
      console.log('   WebSocket Ready:', health.websocket_ready);
      console.log('   Version:', health.version);
      console.log('   Uptime:', Math.floor(health.uptime), 's');

      this.serverVersion = health.version;
      this.serverCapabilities = health.capabilities || [];

      return health.websocket_ready &&
             (health.status === 'healthy' || health.status === 'degraded');

    } catch (error) {
      console.error('❌ Health check error:', error);
      return false;
    }
  }

  async connectToWebSocket() {
    return new Promise((resolve, reject) => {
      const wsURL = this.configuration.websocket.url;
      console.log('🔌 Connecting to WebSocket:', wsURL);

      try {
        this.websocket = new WebSocket(wsURL);

        this.websocket.onopen = () => {
          console.log('✅ WebSocket connected successfully');
          this.setState('connected');
          this.reconnectAttempt = 0;

          // Send capability negotiation handshake
          this.sendHandshake();

          // Send queued messages
          this.flushMessageQueue();

          this.emit('connected');
          resolve();
        };

        this.websocket.onmessage = (event) => {
          try {
            const data = JSON.parse(event.data);
            this.handleMessage(data);
          } catch (error) {
            console.error('❌ Failed to parse WebSocket message:', error);
          }
        };

        this.websocket.onerror = (error) => {
          console.error('❌ WebSocket error:', error);
          this.emit('error', 'WebSocket connection error');
        };

        this.websocket.onclose = (event) => {
          console.log('🔌 WebSocket disconnected', event.code, event.reason);
          this.websocket = null;
          this.setState('disconnected');
          this.emit('disconnected');

          // Auto-reconnect unless it was a normal closure
          if (event.code !== 1000) {
            this.scheduleReconnect();
          }
        };

      } catch (error) {
        console.error('❌ Failed to create WebSocket:', error);
        reject(error);
      }
    });
  }

  sendHandshake() {
    const handshake = {
      type: 'client_connect',
      client_id: this.clientId,
      client_type: this.clientType,
      version: '2.0.0',
      capabilities: this.clientCapabilities,
      request_state: true
    };

    this.send(handshake);
    console.log('📤 Sent capability negotiation handshake');
  }

  handleMessage(data) {
    // Detect welcome message
    if (data.type === 'welcome') {
      console.log('📨 Received welcome message');
      console.log('   Server version:', data.server_version);
      console.log('   Protocol version:', data.protocol_version);
      console.log('   Buffered messages:', data.buffered_messages_count);

      this.serverVersion = data.server_version;
      this.serverCapabilities = data.capabilities || [];
    }

    // Forward to event handlers
    this.emit('message', data);
  }

  async scheduleReconnect() {
    if (!this.configuration || !this.configuration.websocket.reconnect.enabled) {
      console.log('❌ Reconnection disabled');
      return;
    }

    this.reconnectAttempt++;

    if (this.reconnectAttempt > this.configuration.websocket.reconnect.max_attempts) {
      console.error('❌ Max reconnection attempts reached');
      this.emit('error', 'Max reconnection attempts reached');
      return;
    }

    this.setState('reconnecting', { attempt: this.reconnectAttempt });

    // Calculate backoff delay
    const baseDelay = this.configuration.websocket.reconnect.base_delay_ms;
    const maxDelay = this.configuration.websocket.reconnect.max_delay_ms;

    let delay = Math.min(
      baseDelay * Math.pow(2, this.reconnectAttempt - 1),
      maxDelay
    );

    // Add jitter
    if (this.configuration.websocket.reconnect.jitter) {
      const jitter = (Math.random() - 0.5) * 0.2 * delay;
      delay += jitter;
    }

    console.log(`🔄 Reconnecting in ${(delay / 1000).toFixed(1)}s (attempt ${this.reconnectAttempt}/${this.configuration.websocket.reconnect.max_attempts})`);

    this.reconnectTimer = setTimeout(() => {
      this.connect();
    }, delay);
  }

  // MARK: - Helper Methods

  generateClientId() {
    return 'web-' + Math.random().toString(36).substring(2, 15);
  }

  detectBaseURL() {
    // Try to detect from current page
    if (typeof window !== 'undefined') {
      const protocol = window.location.protocol;
      const host = window.location.hostname;
      const port = '8010'; // Default backend port

      return `${protocol}//${host}:${port}`;
    }

    return 'http://localhost:8010';
  }

  setState(state, data = {}) {
    this.connectionState = state;
    this.emit('stateChange', { state, ...data });
  }

  emit(event, data) {
    if (this.eventHandlers[event]) {
      this.eventHandlers[event].forEach(handler => {
        try {
          handler(data);
        } catch (error) {
          console.error(`❌ Event handler error for ${event}:`, error);
        }
      });
    }
  }

  flushMessageQueue() {
    if (this.messageQueue.length > 0) {
      console.log(`📤 Sending ${this.messageQueue.length} queued messages`);
      this.messageQueue.forEach(message => this.send(message));
      this.messageQueue = [];
    }
  }
}

export default UniversalWebSocketClient;
